boards: arm: Add board definition for Adafruit Feather nrf52840 Sense

bkietz opened the pull request:
This is a draft PR since I'm sure I've done something wrong (I can't get blinky or hello_world working locally) and a PR seemed like the most straightforward way to present what I've attempted.

I've added support for adafruit_feather_nrf52840_sense board by copying adafruit_feather_nrf52840 (Express variant) configuration added in https://github.com/zephyrproject-rtos/zephyr/pull/23471 since these boards are similarly constructed.

Apart from trivial updates like names and the image in doc I've made a few changes based on some previous discussion in https://github.com/zephyrproject-rtos/zephyr/issues/32279
- default to UF2 bootloader/runner since the feather sense lacks an SWD header
- modified clock source
- remapped `led0` alias

bkietz committed:
scripts: fix uf2conv.py's whitespace regex

Python 3.12 raises a warning on `'\s+'` since it looks
like an escape.

Signed-off-by: Benjamin Kietzman <bengilgit@gmail.com>

bkietz committed:
doc: delete unused/misplaced link

The link to adafruit pinouts in the doc for arduino
nano 33 ble was probably copypasta. It probably hasn't
been noticed so far because it's not referenced in that
file.

Signed-off-by: Benjamin Kietzman <bengilgit@gmail.com>

bkietz committed:
boards: arm: Add board definition for Adafruit Feather nrf52840 Sense

Add support for adafruit_feather_nrf52840_sense board; mostly
copies the adafruit_feather_nrf52840 (Express variant) configuration
since these boards are similarly constructed.

Signed-off-by: Benjamin Kietzman <bengilgit@gmail.com>

bkietz commented on boards/arm/adafruit_feather_nrf52840_sense/Kconfig:
```diff
@@ -0,0 +1,9 @@
+# Adafruit Feather nRF52840 Sense board configuration
+
```
I wasn't sure what to do with the [copyright notices](https://www.linuxfoundation.org/blog/blog/copyright-notices-in-open-source-software-projects), so I mostly deleted them. I'll be happy to add them back if that was incorrect

bkietz commented on boards/arm/adafruit_feather_nrf52840_sense/adafruit_feather_nrf52840_sense.dts:
```diff
@@ -0,0 +1,167 @@
+
+/ {
+	model = "Adafruit Feather nRF52840 Sense";
+	compatible = "adafruit,feather-nrf52840";
```
I'm not sure what this string does. Should this be:
```suggestion
	compatible = "adafruit,feather-nrf52840-sense";
```

bkietz commented on boards/arm/adafruit_feather_nrf52840_sense/doc/index.rst:
```diff
@@ -0,0 +1,156 @@
+* :abbr:`NVIC (Nested Vectored Interrupt Controller)`
+* :abbr:`PWM (Pulse Width Modulation)`
+* RADIO (Bluetooth Low Energy and 802.15.4)
+* :abbr:`RTC (nRF RTC System Clock)`
```
Is this still accurate since the clock source is different?

bkietz commented on boards/arm/adafruit_feather_nrf52840_sense/doc/index.rst:
```diff
@@ -0,0 +1,156 @@
+
+#. Tap the reset button twice quickly to enter bootloader mode
+
+#. Flash the image.
```
`west build` succeeds, but `west flash` fails with
```
-- west flash: using runner bossac
No device found on /dev/ttyACM0
FATAL ERROR: command exited with status 1: /home/ben/zephyr-sdk-0.16.3/sysroots/x86_64-pokysdk-linux/usr/bin/bossac -p /dev/ttyACM0 -R -e -w -v -b /home/ben/zephyrproject/zephyr/build/zephyr/zephyr.bin -o 49152
```

I'm not familiar with BOSSA; is it not capable of flashing without a serial adapter in addition to the USB cable? I note that the Adafruit ItsyBitsy docs describe connecting both the USB cable and a serial adapter.

bkietz replied:
I've gotten closest to success building a UF2 file and copying that to the bootloader mode mass storage device: the mass storage device unmounts as normal, but the board is not responsive

```
$ west build -p always -b adafruit_feather_nrf52840_sense samples/hello_world -- -DCONFIG_BUILD_OUTPUT_UF2=y -DCONFIG_USB_CDC_ACM=y
$ cp build/zephyr/zephyr.uf2 /media/ben/FTHRSNSBOOT
$ arduino-cli monitor -p /dev/ttyACM0 
Port monitor error: command 'open' failed: no such file or directory
$ sudo dmesg | tail -n 11
[172324.726655] sd 0:0:0:0: [sda] Attached SCSI removable disk
[172334.342045] usb 1-1: USB disconnect, device number 98
[172334.346852] device offline error, dev sda, sector 15 op 0x1:(WRITE) flags 0x0 phys_seg 1 prio class 3
[172334.346885] Buffer I/O error on dev sda, logical block 15, lost async page write
[172334.346932] device offline error, dev sda, sector 273 op 0x1:(WRITE) flags 0x0 phys_seg 1 prio class 3
[172334.346934] Buffer I/O error on dev sda, logical block 273, lost async page write
[172334.346965] device offline error, dev sda, sector 517 op 0x1:(WRITE) flags 0x0 phys_seg 1 prio class 3
[172334.346970] Buffer I/O error on dev sda, logical block 517, lost async page write
[172334.450931] cp: attempt to access beyond end of device
                sda: rw=0, sector=0, nr_sectors = 1 limit=0
[172334.450937] FAT-fs (sda): unable to read boot sector to mark fs as dirty
```

(where the first dmesg line "Attached SCSI removable disk" corresponds to mounting the bootloader mode mass storage device.)

FWIW, I *can* use this method to upload a binary built by arduino-IDE:

```
$ python ~/zephyrproject/zephyr/scripts/build/uf2conv.py -c -f 0xada52840 -b 0xc000 -o ~/Test.uf2 /tmp/arduino/sketches/A186DE617F0051D2152DF2EDCA6EAEC3/Test.ino.hex
$ cp ~/Test.uf2 /media/ben/FTHRSNSBOOT
$ arduino-cli monitor -p /dev/ttyACM0
working...
```

bkietz commented:
Thanks, didn't know that was an option. I'll amend the commit.

bkietz commented:
<details>
<summary>
Further potentially useful info: The arduino definitions for these boards are <a href="https://github.com/adafruit/Adafruit_nRF52_Arduino/tree/1.5.0/variants/feather_nrf52840_express">Express</a> and <a href="https://github.com/adafruit/Adafruit_nRF52_Arduino/tree/1.5.0/variants/feather_nrf52840_sense">Sense</a>, with diff
</summary>

<pre>
diff --git a/.arduino15/packages/adafruit/hardware/nrf52/1.5.0/variants/feather_nrf52840_express/variant.cpp b/.arduino15/packages/adafruit/hardware/nrf52/1.5.0/variants/feather_nrf52840_sense/variant.cpp
index d0ef409..82163aa 100644
--- a/.arduino15/packages/adafruit/hardware/nrf52/1.5.0/variants/feather_nrf52840_express/variant.cpp
+++ b/.arduino15/packages/adafruit/hardware/nrf52/1.5.0/variants/feather_nrf52840_sense/variant.cpp
@@ -29,7 +29,7 @@ const uint32_t g_ADigitalPinMap[] =
   25,  // D0  is P0.25 (UART TX)
   24,  // D1  is P0.24 (UART RX 
   10,  // D2  is P0.10 (NFC2)
-  47,  // D3  is P1.15 (LED1)
+  43,  // D3  is P1.11
   42,  // D4  is P1.10 (LED2)
   40,  // D5  is P1.08
    7,  // D6  is P0.07
@@ -39,7 +39,7 @@ const uint32_t g_ADigitalPinMap[] =
   27,  // D10 is P0.27
    6,  // D11 is P0.06
    8,  // D12 is P0.08
-  41,  // D13 is P1.09
+  41,  // D13 is P1.09 (LED1)
 
   // D14 .. D21 (aka A0 .. A7)
    4,  // D14 is P0.04 (A0)
@@ -69,18 +69,15 @@ const uint32_t g_ADigitalPinMap[] =
   21,  // D32 is P0.21 (QSPI Data 3)
 
   // The remaining NFC pin
-   9,  // D33 is P0.09 (NFC1, exposed only via test point on bottom of board)
+   9,   // D33 is P0.09 (NFC1, exposed only via test point on bottom of board)
 
-  // Thus, there are 34 defined pins
+   0,   // D34 P0.00 is PDM Data
+   1,   // D35 P0.01 is PDM Clock
+  32,   // D36 P1.00 is APDS IRQ
 
   // The remaining pins are not usable:
-  //
-  //
   // The following pins were never listed as they were considered unusable
-  //  0,      // P0.00 is XL1   (attached to 32.768kHz crystal)
-  //  1,      // P0.01 is XL2   (attached to 32.768kHz crystal)
   // 18,      // P0.18 is RESET (attached to switch)
-  // 32,      // P1.00 is SWO   (attached to debug header)
   // 
   // The remaining pins are not connected (per schematic)
   // 33,      // P1.01 is not connected per schematic
diff --git a/.arduino15/packages/adafruit/hardware/nrf52/1.5.0/variants/feather_nrf52840_express/variant.h b/.arduino15/packages/adafruit/hardware/nrf52/1.5.0/variants/feather_nrf52840_sense/variant.h
index f112e0f..210f09a 100644
--- a/.arduino15/packages/adafruit/hardware/nrf52/1.5.0/variants/feather_nrf52840_express/variant.h
+++ b/.arduino15/packages/adafruit/hardware/nrf52/1.5.0/variants/feather_nrf52840_sense/variant.h
@@ -16,14 +16,13 @@
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
 
-#ifndef _VARIANT_FEATHER52840_
-#define _VARIANT_FEATHER52840_
+#ifndef _VARIANT_FEATHER52840_SENSE_
+#define _VARIANT_FEATHER52840_SENSE_
 
 /** Master clock frequency */
 #define VARIANT_MCK       (64000000ul)
 
-#define USE_LFXO      // Board uses 32khz crystal for LF
-// define USE_LFRC    // Board uses RC for LF
+#define USE_LFRC    // Board uses RC for LF
 
 /*----------------------------------------------------------------------------
  *        Headers
@@ -37,13 +36,13 @@ extern "C"
 #endif // __cplusplus
 
 // Number of pins defined in PinDescription array
-#define PINS_COUNT           (34)
-#define NUM_DIGITAL_PINS     (34)
+#define PINS_COUNT           (37)
+#define NUM_DIGITAL_PINS     (37)
 #define NUM_ANALOG_INPUTS    (6) // A6 is used for battery, A7 is analog reference
 #define NUM_ANALOG_OUTPUTS   (0)
 
 // LEDs
-#define PIN_LED1             (3)
+#define PIN_LED1             (13)
 #define PIN_LED2             (4)
 #define PIN_NEOPIXEL         (8)
 #define NEOPIXEL_NUM         1
@@ -59,7 +58,12 @@ extern "C"
 /*
  * Buttons
  */
-#define PIN_BUTTON1          (7)
+#define PIN_BUTTON1             (7)
+
+// Microphone
+#define PIN_PDM_DIN           34
+#define PIN_PDM_CLK           35
+#define PIN_PDM_PWR           -1  // not used
 
 /*
  * Analog pins
</pre>
</details>

bkietz committed:
boards: arm: Refactor Feather nrf52840 Sense to an extension of Express

This reduces duplication of between the Express and Sense definitions
significantly.

Signed-off-by: Benjamin Kietzman <bengilgit@gmail.com>

bkietz commented on boards/arm/adafruit_feather_nrf52840/doc/index.rst:
```diff
@@ -103,33 +132,72 @@ Push buttons
+is compatible with BOSSA and supports flashing using `UF2`_. This allows
+easy flashing of new images, but does not support debugging the device.
+
+Additionally, if :code:`CONFIG_USB_CDC_ACM` is enabled then the BOSSA
```
Could/should this be added to `_defconfig`s?

bkietz commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_sense.yaml:
```diff
@@ -0,0 +1,19 @@
+identifier: adafruit_feather_nrf52840_sense
```
@kartben this identifier isn't picked up by `CONFIG_BOARD`, so hello_world would print "adafruit_feather_nrf52840" even when built against "adafruit_feather_nrf52840_sense". This isn't a huge problem, but is there another macro one could examine to determine in C which variant was built?

bkietz replied:
Maybe this would be a job for `BOARD_REVISION`? However the `xiao_ble` doesn't use that

bkietz replied:
I'd expect the `model` field from adafruit_feather_nrf52840_sense.dts to do the job, but I don't see it in devicetree_generated.h

kartben replied:
This identifier is not the one used at runtime, this is how you would typically do it https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/rpi_pico/Kconfig.defconfig#L6-L8. 
It looks like xiao_ble doesn't do things quite right (I am quite certain that hello world on XIAO BLE Sense would say "xiao_ble" and you might be better off looking at pico/picow.

bkietz replied:
That did it; `strings build/zephyr/zephyr.bin | rg adafruit` now shows the _sense suffix. Thanks!

bkietz committed:
boards: arm: Ensure distinct CONFIG_BOARD for Express and Sense

Previously CONFIG_BOARD would be "adafruit_feather_nrf52840"
even when building against "adafruit_feather_nrf52840_sense".

Signed-off-by: Benjamin Kietzman <bengilgit@gmail.com>

bkietz commented on boards/arm/adafruit_feather_nrf52840/Kconfig:
```diff
@@ -7,4 +7,5 @@ config BOARD_ENABLE_DCDC
 	select SOC_DCDC_NRF52X
 	default y
-	depends on BOARD_ADAFRUIT_FEATHER_NRF52840
+	depends on BOARD_ADAFRUIT_FEATHER_NRF52840 \
+		|| BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE
```
Should I replace this `depends on` with an `if` block [as in the xiao_ble](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/xiao_ble/Kconfig#L6C1)? The rpi_pico doesn't seem to have a corresponding option (or a Kconfig file)

kartben commented on boards/arm/adafruit_feather_nrf52840/Kconfig.defconfig:
you will probably want to factor things out, similar to ex. KConfig.defconfig for rpi_pico, as this will lead to less duplication / better readability.

bkietz replied:
I've moved the contents of Kconfig to Kconfig.defconfig as with the rpi_pico

nordicjm commented on boards/arm/adafruit_feather_nrf52840/Kconfig.defconfig:
```diff
@@ -8,7 +8,27 @@ if BOARD_ADAFRUIT_FEATHER_NRF52840
 	default BT
 
-endif # BOARD_ADAFRUIT_FEATHER_NRF52840
+config BOARD_ENABLE_DCDC
+	bool "DCDC mode"
+	select SOC_DCDC_NRF52X
+	default y
```
No this goes in the Kconfig file

nordicjm commented on boards/arm/adafruit_feather_nrf52840/Kconfig:
```diff
@@ -1,10 +0,0 @@
-# Copyright (c) 2020 Tobias Svehagen
-# SPDX-License-Identifier: Apache-2.0
-
-config BOARD_ENABLE_DCDC
```
i.e. restore this

bkietz replied:
I moved this on the advice of @kartben, could you point me to doc describing the differences betwen Kconfig and Kconfig.defconfig?  I didn't see any guidance [here](https://docs.zephyrproject.org/latest/build/kconfig/setting.html#visible-and-invisible-kconfig-symbols) for example

nordicjm replied:
`Kconfig.defconfig` is strictly for overriding defaults in zephyr, `Kconfig` is for options that a user can select. @kartben take note

kartben replied:
FWIW I don't remember suggesting this change so there might have been a misunderstanding :)

bkietz replied:
Will revert

nordicjm commented on boards/arm/adafruit_feather_nrf52840/Kconfig.defconfig:
```diff
@@ -8,7 +8,27 @@ if BOARD_ADAFRUIT_FEATHER_NRF52840
 config BOARD
 	default "adafruit_feather_nrf52840"
 
+endif # BOARD_ADAFRUIT_FEATHER_NRF52840
```
```suggestion
if(BOARD_ADAFRUIT_FEATHER_NRF52840 || BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE)

config BOARD
	default "adafruit_feather_nrf52840" if BOARD_ADAFRUIT_FEATHER_NRF52840
	default "adafruit_feather_nrf52840_sense" if BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE

endif()
```

nordicjm commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_sense.dts:
```diff
@@ -0,0 +1,167 @@
+	compatible = "adafruit,feather-nrf52840";
+
+	chosen {
+		zephyr,console = &uart0;
+		zephyr,shell-uart = &uart0;
+		zephyr,uart-mcumgr = &uart0;
+		zephyr,bt-mon-uart = &uart0;
+		zephyr,bt-c2h-uart = &uart0;
```
Should this be UART, or should this be USB CDC?

bkietz replied:
I copied most of this from the Express variant. Since I haven't been able to flash the board (Sense) that I have, most of these options are opaque to me. I posted this PR to more articulately ask for help from people who *have* had success like @jefffhaynes. I'd also love to hear from @tsvehagen who originally added support for the Express board.

nordicjm replied:
> Since I haven't been able to flash the board (Sense) that I have

What is not working, surely the device should just load the uf2 file you have generated? Or do you mean that it does not run it?

bkietz replied:
It does not run it: https://github.com/zephyrproject-rtos/zephyr/pull/63819#discussion_r1355416877

nordicjm replied:
I would assume that the bootloader is not compatible with bossac, would make more sense to support the uf2 runner. I've flashed their bootloader to a different board here and can confirm that the uf2 flash runner works though (in linux) the partition must be mounted for it to detect it

bkietz replied:
Didn't see there was a runner for UF2. I've made that the default runner and removed BOSSA.

I still can't flash blinky however. `$ west flash` fails with:
<details>
<summary>(click for errors)</summary>
<pre>
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner uf2
-- runners.uf2: Copying UF2 file to '/media/ben/FTHRSNSBOOT'
Traceback (most recent call last):
  File "/home/ben/mambaforge/envs/zephyr/bin/west", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/ben/mambaforge/envs/zephyr/lib/python3.12/site-packages/west/app/main.py", line 10
85, in main
    app.run(argv or sys.argv[1:])
  File "/home/ben/mambaforge/envs/zephyr/lib/python3.12/site-packages/west/app/main.py", line 24
4, in run
    self.run_command(argv, early_args)
  File "/home/ben/mambaforge/envs/zephyr/lib/python3.12/site-packages/west/app/main.py", line 50
5, in run_command
    self.run_extension(args.command, argv)
  File "/home/ben/mambaforge/envs/zephyr/lib/python3.12/site-packages/west/app/main.py", line 65
4, in run_extension
    self.cmd.run(args, unknown, self.topdir, manifest=self.manifest,
  File "/home/ben/mambaforge/envs/zephyr/lib/python3.12/site-packages/west/commands.py", line 19
4, in run
    self.do_run(args, unknown)
  File "/home/ben/zephyrproject/zephyr/scripts/west_commands/flash.py", line 33, in do_run
    do_run_common(self, my_args, runner_args, domains=domains)
  File "/home/ben/zephyrproject/zephyr/scripts/west_commands/run_common.py", line 174, in do_run
_common
    do_run_common_image(command, user_args, user_runner_args, d.build_dir)
  File "/home/ben/zephyrproject/zephyr/scripts/west_commands/run_common.py", line 250, in do_run
_common_image
    runner.run(command_name)
  File "/home/ben/zephyrproject/zephyr/scripts/west_commands/runners/core.py", line 621, in run
    self.do_run(command, **kwargs)
  File "/home/ben/zephyrproject/zephyr/scripts/west_commands/runners/uf2.py", line 105, in do_ru
n
    self.copy_uf2_to_partition(part)
  File "/home/ben/zephyrproject/zephyr/scripts/west_commands/runners/uf2.py", line 88, in copy_u
f2_to_partition
    copy(self.cfg.uf2_file, part.mountpoint)
  File "/home/ben/mambaforge/envs/zephyr/lib/python3.12/shutil.py", line 424, in copy
    copymode(src, dst, follow_symlinks=follow_symlinks)
  File "/home/ben/mambaforge/envs/zephyr/lib/python3.12/shutil.py", line 313, in copymode
    chmod_func(dst, stat.S_IMODE(st.st_mode))
FileNotFoundError: [Errno 2] No such file or directory: '/media/ben/FTHRSNSBOOT/zephyr.uf2'
</pre>
</details>

nordicjm replied:
You will need to use python 3.11, python 3.12 is not supported

bkietz replied:
I didn't see that 3.12 is not supported. I've replaced my environment with on using 3.11, and I still get a very similar error. I've updated the above traceback just in case I missed something

nordicjm replied:
So that error looks like you do not have permission to copy files to that drive, can you manually copy the file and does it work?

bkietz replied:
Yes, as I described [above](https://github.com/zephyrproject-rtos/zephyr/pull/63819#discussion_r1355416877) I can copy UF2 files produced with arduino IDE onto the drive. Also manually copying the UF2 file produced by west does not raise an error (`cp` exits with code 0).

nordicjm commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_sense.dts:
```diff
@@ -0,0 +1,167 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		boot_partition: partition@0 {
+			label = "mcuboot";
+			reg = <0x00000000 0x0000C000>;
+		};
+		slot0_partition: partition@c000 {
+			label = "image-0";
+			reg = <0x0000C000 0x00067000>;
+		};
+		slot1_partition: partition@73000 {
+			label = "image-1";
+			reg = <0x00073000 0x00067000>;
+		};
+		scratch_partition: partition@da000 {
+			label = "image-scratch";
+			reg = <0x000da000 0x0001e000>;
+		};
```
This doesn't look right, would someone load mcuboot to this board? If so, how can they load it to 0x0 when there is already some sort of bootloader on it given it accepts UF2 files?

nordicjm replied:
Here is a sample that will likely work on your board:
```
		mbr_partition: partition@0 {
			label = "mbr";
			read-only;
			reg = <0x00000000 0x00001000>;
		};
		boot_partition: partition@1000 {
			label = "mcuboot";
			reg = <0x00001000 0x0000C000>;
		};
		slot0_partition: partition@d000 {
			label = "image-0";
			reg = <0x0000d000 0x00070000>;
		};
		slot1_partition: partition@7d000 {
			label = "image-1";
			reg = <0x0007d000 0x0006f000>;
		};
		storage_partition: partition@ec000 {
			label = "storage";
			reg = <0x000ec000 0x00008000>;
		};
		uf2_bootloader_partition: partition@f4000 {
			label = "uf2-bootloader";
			reg = <0x000f4000 0x0000c000>;
		};
```
Though would require mcuboot. Here it is without mcuboot support:
```
		mbr_partition: partition@0 {
			label = "mbr";
			read-only;
			reg = <0x00000000 0x00001000>;
		};
		slot0_partition: partition@1000 {
			label = "image-0";
			reg = <0x00001000 0x000eb000>;
		};
		storage_partition: partition@ec000 {
			label = "storage";
			reg = <0x000ec000 0x00008000>;
		};
		uf2_bootloader_partition: partition@f4000 {
			label = "uf2-bootloader";
			reg = <0x000f4000 0x0000c000>;
		};
```

nordicjm commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_sense.dts:
```diff
@@ -0,0 +1,167 @@
+			label = "image-1";
+			reg = <0x00073000 0x00067000>;
+		};
+		scratch_partition: partition@da000 {
+			label = "image-scratch";
+			reg = <0x000da000 0x0001e000>;
+		};
```
And no scratch

nordicjm commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_sense.dts:
```diff
@@ -0,0 +1,167 @@
+	pinctrl-0 = <&qspi_default>;
+	pinctrl-1 = <&qspi_sleep>;
+	pinctrl-names = "default", "sleep";
+	gd25q16: gd25q16@0 {
```
I cannot see any mention of this on the product page?

nordicjm commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_sense.dts:
```diff
@@ -0,0 +1,167 @@
+	pinctrl-names = "default", "sleep";
+};
+
+&i2c0 {
+	compatible = "nordic,nrf-twi";
+	status = "okay";
+	pinctrl-0 = <&i2c0_default>;
+	pinctrl-1 = <&i2c0_sleep>;
+	pinctrl-names = "default", "sleep";
+};
+
+&spi1 {
+	compatible = "nordic,nrf-spi";
+	status = "okay";
+	pinctrl-0 = <&spi1_default>;
+	pinctrl-1 = <&spi1_sleep>;
+	pinctrl-names = "default", "sleep";
+};
```
Surely this board should have some sensors...?

nordicjm commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_sense.yaml:
```diff
@@ -0,0 +1,19 @@
+name: Adafruit Feather nRF52840 Sense
+type: mcu
+arch: arm
+toolchain:
```
```
ram: 256
flash: 1024
```

nordicjm commented on boards/arm/adafruit_feather_nrf52840/board.cmake:
```diff
@@ -2,6 +2,7 @@
+include(${ZEPHYR_BASE}/boards/common/bossac.board.cmake)
 include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
 include(${ZEPHYR_BASE}/boards/common/pyocd.board.cmake)
 include(${ZEPHYR_BASE}/boards/common/blackmagicprobe.board.cmake)
```
264KB image? Please scale that way down to 50KB or so by cropping it and using https://tinypng.com/

bkietz replied:
Done

nordicjm commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_sense_defconfig:
```diff
@@ -0,0 +1,30 @@
+
+# Build UF2 by default, supported by the Adafruit nRF52 Bootloader
+CONFIG_BUILD_OUTPUT_UF2=y
+CONFIG_USE_DT_CODE_PARTITION=y
```
According to https://docs.zephyrproject.org/latest/develop/flash_debug/host-tools.html this final line should not be needed

> 2. Have board :file:`.defconfig` file with the :kconfig:option:`CONFIG_BOOTLOADER_BOSSA` Kconfig option set to ``y``.  This will automatically select the :kconfig:option:`CONFIG_USE_DT_CODE_PARTITION` Kconfig option which instruct the build system to use these partitions for code relocation.  The board :file:`.defconfig` file should have :kconfig:option:`CONFIG_BOOTLOADER_BOSSA_ARDUINO` , :kconfig:option:`CONFIG_BOOTLOADER_BOSSA_ADAFRUIT_UF2` or the :kconfig:option:`CONFIG_BOOTLOADER_BOSSA_LEGACY` Kconfig option set to ``y`` to select the right compatible SAM-BA bootloader mode. These options can also be set in ``prj.conf`` or any other Kconfig fragment.

bkietz replied:
I'll remove it, thanks

bkietz replied:
should I also remove it from adafruit_itsybitsy_nrf52840_defconfig?

nordicjm replied:
Can do, but do that in a separate PR

nordicjm commented on boards/arm/adafruit_feather_nrf52840/adafruit_feather_nrf52840_defconfig:
```diff
@@ -16,3 +16,11 @@ CONFIG_SERIAL=y
+
+# Build UF2 by default, supported by the Adafruit nRF52 Bootloader
+CONFIG_BUILD_OUTPUT_UF2=y
+CONFIG_USE_DT_CODE_PARTITION=y
```
As below

nordicjm commented on boards/arm/adafruit_feather_nrf52840/doc/index.rst:
```diff
@@ -103,33 +132,72 @@ Push buttons
+easy flashing of new images, but does not support debugging the device.
+
+Additionally, if :code:`CONFIG_USB_CDC_ACM` is enabled then the BOSSA
+bootloader will be entered automatically when you run :code:`west flash`.
```
I don't have one of these boards to test but is this really true?

bkietz replied:
This is again copied from another board without a lot of understanding, in this case the [Adafruit ItsyBitsy M4 Express](https://github.com/bkietz/zephyr/blob/5bc208e16c7ae6f94e56efa5fd5be6829b8dc35e/boards/arm/adafruit_itsybitsy_m4_express/doc/index.rst?plain=1#L112-L120)

bkietz replied:
I *can* verify that with these options `west flash` invokes BOSSA

nordicjm replied:
Invokes sure, but if you have a board running a zephyr application and run west flash, it somehow reboots it into bootloader mode?

bkietz replied:
I haven't managed a zephyr application yet, but it does reboot it into bootloader mode from a running arduino application

nordicjm commented on boards/arm/adafruit_feather_nrf52840/doc/index.rst:
```diff
@@ -103,33 +132,72 @@ Push buttons
+         :zephyr-app: samples/basic/blinky
+         :board: adafruit_feather_nrf52840
+         :goals: build
+         :gen-args: -DCONFIG_USB_CDC_ACM=y
```
Why bother with this? Just make USB CDC the default console and tty device in the dts, doesn't really make sense to use a UART that someone would need to solder to some other UART interface port to actually use

bkietz replied:
will do

bkietz replied:
I tried to change the default in https://github.com/zephyrproject-rtos/zephyr/pull/63819/commits/4964a437d5bcd21224b78679a5ecee3cf7924a4e but this has introduced an error I don't know how to address:

```
gen_isr_tables.py: error: multiple registrations at table_index 39 for irq 39 (0x27)
Existing handler 0x1eb59, new handler 0x1eb59
Has IRQ_CONNECT or IRQ_DIRECT_CONNECT accidentally been invoked on the same irq multiple times?
```

nordicjm replied:
First thing is get rid of the label

bkietz committed:
boards: arm: Make USB CDC the default console for Feather nrf52840

Per review comment: since usb is always available it seems preferable
as the default console (on the Sense variant using UART requires
soldering).

DeviceTree and Kconfig changes were modeled on
`adafruit_itsybitsy_nrf52840`'s board definition.

Signed-off-by: Benjamin Kietzman <bengilgit@gmail.com>

bkietz committed:
boards: arm: Use UF2 runner as the default for Feather-nrf52840

Board IDs and the UF2 volume labels were drawn from
https://github.com/adafruit/Adafruit_nRF52_Bootloader/
- Express: src/boards/feather_nrf52840_express/board.h
- Sense: src/boards/feather_nrf52840_sense/board.h

Signed-off-by: Benjamin Kietzman <bengilgit@gmail.com>

bkietz committed:
boards: arm: Move shared DeviceTree into -common.dtsi

(Modulo the sensors, which haven't been added yet)
the boards only differ in their clock sources and which
GPIO pin is the red LED.

Signed-off-by: Benjamin Kietzman <bengilgit@gmail.com>
